packages/backend/src/note/routes.rs

```diff
@@ -80,11 +81,20 @@ pub async fn create(Json(mut n): Json<Note>) -> Response {
     }
 }
 
-pub async fn delete(Path(OneNoteParams { id }): Path<OneNoteParams>) -> Response {
+pub async fn delete(
+    Path(OneNoteParams { id }): Path<OneNoteParams>,
+    state: axum::extract::State<SharedState>,
+) -> Response {
+    let mut locks_map = state.locks.lock().await;
+    let lock = locks_map
+        .entry(id.clone())
+        .or_insert_with(|| Arc::new(Mutex::new(())))
+        .clone();
+    drop(locks_map);
+    let _guard = lock.lock().await;
```

Comment:
Correct me if I'm wrong but this will provide in-process concurrency protection but is not distributed. If the backend were deployed in a multi-instance environment, I believe the race condition still exists. This is probably fine for most deployments but is a notable consideration.

It may be worth refactoring to have the read & delete operation contained in a single store method. i.e. 'read' knows when to also delete. This control shift allows for the store itself to perform read & delete operations atomically for the respective storage implementation. In this case, a Redis lua script could be used.
Alternatively, a more generalized distributed locking mechanism could be used to wrap the read and delete instead of the rust Arc one.

Reply:
You are right, this guarantees locking within one running instance of the backend.
[One could do locking IN redis](https://redis.io/docs/latest/develop/use/patterns/distributed-locks/#disclaimer-about-consistency). I will skip for now i think, but will document in the readme.
Very good point and spot on!